core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -126,6 +126,9 @@
     @GuardedBy("this")
     private ListenableFuture<Void> blocked = immediateVoidFuture();
 
+    @GuardedBy("this")
+    private ListenableFuture<Void> tasksPopulatedFuture = immediateVoidFuture();
```

Comment:
Sorry, I totally forgot how I discovered this bug. It's been a while ago. I forgot to separate into a separate commit one more thing.

In this PR I'm also changing when a stage is considered `FINISHED`. Originally a stage would be transitioned to `FINISHED` only after query completion. It was done to allow tasks to be re-scheduled even when all the tasks from a given stage succeeded (this was needed to allow re-computation when the data is lost). This mechanism has never been implemented and it's unclear if it will ever be. Thus I decided to close a stage early while refactoring (Let me extract this into a separate commit).

When I tried that I saw that a stage is getting transitioned to `FINISHED` when there are still tasks to be run. That's how I discovered a bug that a fault tolerant scheduler may report being finished while still have some work pending. I will add a regression test for it as well.